### PR TITLE
Add memory store/retrieve tools with API and tests

### PR DESCRIPTION
**Pull request by leonj1**
## Summary
Introduce a Memory Management feature for Archon MCP Server to store and retrieve contextual memories across sessions. Includes a dedicated API, MCP tools, a SQLite-backed data store, and unit tests. This complements existing retrieval (RAG) by persisting patterns, solutions, API discoveries, and learnings.

## Changes
- Backend API
  - Added memory API endpoints: 
    - POST /api/memory/store
    - GET /api/memory/retrieve/{memory_key}
    - POST /api/memory/search
    - GET /api/memory/list
    - DELETE /api/memory/{memory_id}
- MCP Tools
  - Implemented memory management tools in python/src/mcp_server/features/memory/memory_tools.py
  - Tools: `store_memory` and `retrieve_memory` exposed via MCP, with usage tracking
- Memory Service & Data Layer
  - Added python/src/server/services/memory_service.py with:
    - store_memory(memory_key, memory_content, memory_type, session_id, tags, metadata)
    - retrieve_memory_by_key(memory_key)
    - search_memories(query, memory_type, tags, session_id, match_count)
    - list_memories(memory_type, session_id, limit)
    - delete_memory(memory_id)
  - Integrates embedding generation for semantic storage and maintains access stats
- Database Migration
  - New migration script: migration/sqlite/003_memory_store.sql
  - Creates archon_mcp_memories and archon_mcp_memory_stats tables, indexes, and triggers for maintenance
- Server Wiring
  - Memory API router wired into server: python/src/server/main.py
  - MCP server registers memory tools during module registration (optional, robust to absence)
- Documentation
  - New README: python/src/mcp_server/features/memory/README.md
- Tests
  - New unit tests: python/tests/server/services/test_memory_service.py validating store, retrieve, search, list, and delete flows using a temp SQLite DB

## Migration Notes
- Adds persistent storage for memories with the following schema:
  - archon_mcp_memories(id, memory_key, memory_content, memory_type, session_id, tags, metadata, embedding, created_at, updated_at)
  - archon_mcp_memory_stats(id, memory_id, access_count, last_accessed)
- Triggers for updated_at, memory access, and cleanup policies (90 days global, 30 days session-scoped)

## API Surface
- POST /api/memory/store
  - Body: memory_key, memory_content, memory_type, session_id, tags, metadata
- GET /api/memory/retrieve/{memory_key}
  - Retrieve memory by exact key
- POST /api/memory/search
  - Body: query, memory_type, tags, session_id, match_count
- GET /api/memory/list
  - Optional filters: memory_type, session_id, limit
- DELETE /api/memory/{memory_id}
  - Delete memory by ID

## Usage Examples
- Store Memory (via MCP tool)
  - store_memory(
      key="fastapi_jwt_middleware",
      content="Use @app.middleware with request.state for JWT validation; store decoded token in request.state.user.",
      type="pattern",
      tags=["authentication","fastapi","jwt","middleware"]
    )
- Retrieve by Key
  - retrieve_memory(key="fastapi_jwt_middleware")
- Search Memories
  - retrieve_memory(query="authentication", type="pattern", tags=["jwt"], match_count=3)

## Data Model & Retention
- Memory types: pattern, solution, api, architecture, learning (default)
- Global memories: retain 90 days; session-scoped memories: 30 days
- Automatic cleanup handled by database triggers

## Testing Plan
- [x] Unit tests for MemoryService covering store, update, retrieve, search, list, and delete
- [x] Tests use a temporary SQLite database to ensure isolation
- [x] Validate handling of edge cases (empty key/content, non-existent memory)

## How to Run Tests
- cd python
- pytest tests/server/services/test_memory_service.py -v

## Notes
- Memory tools are registered as optional MCP tools; if the memory module is unavailable, the server continues to run with warnings.
- This feature is designed to be extended with richer search (embedding-based) and more granular permissions in future PRs.

🌿 Generated by [Terry](https://www.terragonlabs.com)

---

ℹ️ Tag @terragon-labs to ask questions and address PR feedback

📎 **Task**: https://www.terragonlabs.com/task/462d0bf4-af74-467b-9772-d9681221c1b9

## Summary by CodeRabbit

* **New Features**
  * Persistent Memory Management: store, retrieve, search, list, and delete memories with types, tags, session scope, usage stats, API endpoints, and MCP tooling (find/manage workflows).
  * Automatic retention/cleanup (defaults: 90 days global, 30 days session).

* **Documentation**
  * Added usage guides, API examples, tool descriptions, architecture and operator notes for memory features.

* **Tests**
  * Comprehensive test suite covering store/retrieve/search/list/delete behaviors.

### CODE REVIEW

File: CLAUDE.md [hunk 1]
```diff
@@ -275,6 +275,15 @@ When connected to Claude/Cursor/Windsurf, the following tools are available:
...
+- `archon:store_memory` - Store knowledge, patterns, or learnings for future recall
+  - Parameters: `key`, `content`, `type` (pattern/solution/api/architecture/learning), `tags`, `session_id`, `metadata`
+  - Example: Store authentication patterns, error solutions, API discoveries
+- `archon:retrieve_memory` - Retrieve stored memories by exact key or search
+  - Parameters: `key` (exact match) OR `query/type/tags` (search)
+  - Returns: Single memory (by key) or list of matching memories (search)
+
 ### Project Management
 
 - `archon:find_projects` - Find all projects, search, or get specific project (by project_id)
```
[the remark this answers is not in the document]
Comment (leonj1) [reply CLAUDE.md:290]: @terragon-labs

File: migration/sqlite/003_memory_store.sql [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    memory_key TEXT NOT NULL,
+    memory_content TEXT NOT NULL,
+    memory_type TEXT NOT NULL DEFAULT 'learning',
+    session_id TEXT,
+    tags TEXT,
+    metadata TEXT,
+    embedding BLOB,
+    created_at DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    updated_at DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP
+);
```
[the remark this answers is not in the document]
Comment (leonj1) [reply migration/sqlite/003_memory_store.sql:19]: @terragon-labs

File: python/src/mcp_server/features/memory/memory_tools.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+                        "message": result.get("message"),
+                    }, indent=2)
+                else:
+                    return MCPErrorFormatter.from_http_error(response, "store memory")
+
+        except httpx.RequestError as e:
+            return MCPErrorFormatter.from_exception(e, "store memory")
+        except Exception as e:
+            logger.error(f"Error storing memory: {e}", exc_info=True)
+            return MCPErrorFormatter.from_exception(e, "store memory")
```
[the remark this answers is not in the document]
Comment (leonj1) [reply python/src/mcp_server/features/memory/memory_tools.py]: @terragon-labs

File: python/src/mcp_server/features/memory/memory_tools.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+                            "error": None,
+                        }, indent=2)
+                    else:
+                        return MCPErrorFormatter.from_http_error(response, "search memories")
+
+        except httpx.RequestError as e:
+            return MCPErrorFormatter.from_exception(e, "retrieve memory")
+        except Exception as e:
+            logger.error(f"Error retrieving memory: {e}", exc_info=True)
+            return MCPErrorFormatter.from_exception(e, "retrieve memory")
```
[the remark this answers is not in the document]
Comment (leonj1) [reply python/src/mcp_server/features/memory/memory_tools.py]: @terragon-labs

File: python/src/mcp_server/features/memory/README.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  - `GET /api/memory/retrieve/{key}` - Retrieve by exact key
+  - `POST /api/memory/search` - Search memories
+  - `GET /api/memory/list` - List all memories
+  - `DELETE /api/memory/{id}` - Delete memory
+
+### MCP Tools
+- **Location**: `python/src/mcp_server/features/memory/memory_tools.py`
+- **Tools**:
+  - `store_memory` - Store knowledge for future recall
+  - `retrieve_memory` - Retrieve by key or search
```
[the remark this answers is not in the document]
Comment (leonj1) [reply python/src/mcp_server/features/memory/README.md]: @terragon-labs

File: python/src/mcp_server/mcp_server.py [hunk 6]
```diff
@@ -267,6 +267,26 @@ async def lifespan(server: FastMCP) -> AsyncIterator[ArchonContext]:
...
+- `retrieve_memory(key, query, type, tags, session_id, match_count)`
+  - Retrieve by exact key or search
+  - Example (exact): `retrieve_memory(key="fastapi_jwt_auth")`
+  - Example (search): `retrieve_memory(query="authentication", type="pattern", tags=["jwt"])`
+
+### When to Use Memory
+- **Store** after discovering patterns, solutions, or architectural decisions
+- **Retrieve** before starting new tasks to check for existing knowledge
+- **Tag** memories with relevant keywords for easy discovery
+
```
[the remark this answers is not in the document]
Comment (leonj1) [reply python/src/mcp_server/mcp_server.py:290]: @terragon-labs

File: python/tests/server/services/test_memory_service.py [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+    assert success
+    assert all(m["memory_type"] == "pattern" for m in result["memories"])
+
+
+@pytest.mark.asyncio
+async def test_list_memories_invalid_limit_fails(memory_service):
+    """Test that listing with invalid limit fails."""
+    success, result = await memory_service.list_memories(limit=0)
+
+    assert not success
```
[the remark this answers is not in the document]
Comment (leonj1) [reply python/tests/server/services/test_memory_service.py:424]: tt

[the remark this answers is not in the document]
Comment (leonj1) [reply python/tests/server/services/test_memory_service.py:424]: @terragon-labs
